File: code/cpp/ch25/solutions/challenge_04_sol.cpp
/*
 * Solution for Challenge 4: Longest String Chain
 * Chapter 25: Dynamic Programming III — Subsequences & Knapsack
 */
#include <algorithm>
#include <climits>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

int solve(vector<string>& words) {
    sort(words.begin(), words.end(), [](const string& a, const string& b) {
        return a.size() < b.size();
    });
    unordered_map<string, int> dp;
    int best = 1;
    for (const string& word : words) {
        dp[word] = 1;
        for (int i = 0; i < (int)word.size(); i++) {
            string pred = word.substr(0, i) + word.substr(i + 1);
            if (dp.count(pred))
                dp[word] = max(dp[word], dp[pred] + 1);
        }
        best = max(best, dp[word]);
    }
    return best;
}

int main() { return 0; }

File: code/python/ch25/solutions/challenge_04_sol.py
"""
Solution for Challenge 4: Longest String Chain
=================================================
Chapter 25: Dynamic Programming III — Subsequences & Knapsack

APPROACH
--------
Sort words by length. For each word, try removing each character
to form a predecessor and look it up in a hash map.
dp[word] = length of longest chain ending at word.

TIME COMPLEXITY:  O(n * L^2) where L = max word length
SPACE COMPLEXITY: O(n * L)
"""


def solve(words: list[str]) -> int:
    """Return the length of the longest string chain."""
    words.sort(key=len)
    dp: dict[str, int] = {}
    best = 1
    for word in words:
        dp[word] = 1
        for i in range(len(word)):
            pred = word[:i] + word[i + 1:]
            if pred in dp:
                dp[word] = max(dp[word], dp[pred] + 1)
        best = max(best, dp[word])
    return best


# ── Do not change anything below this line ──────────────────────
if __name__ == "__main__":
    words = input().strip().split()
    print(solve(words))
